계산기 II [STEP 1] 애종, Dragon

jonghancha opened the pull request:
# 계산기 II [STEP 1] 애종, Dragon

@hyunable 
안녕하세요! hyunable
계산기 II - STEP1 완료하고 PR 드립니다.

팀원 각각의 코드를 비교해본 결과 기능적으로 크게 다른 부분이 없어서 수정된 코드는 많지 않았습니다.
큰 차이점으로는 큐 구현을 링크드리스트와 배열로 구현되어있었는데 이 부분은 링크드리스트방식을 사용하기로 하였고, 다른 차이점은 아래 `고민했던 점`에 내용 추가하였습니다!~

---

## 고민했던 점
 ### 1. 에러 케이스 추가
 - Formula.result()메서드 내부의 dequeue()를 옵셔널바인딩하는 과정에서 에러가 발생했을 경우, 0만 반환시키지 않고 `emptyOperands`, `emptyOperators의` 에러타입을 던져주도록 수정했습니다.
 
 ### 2. 고차함수 사용
 - ExpressionParser 에서 배열 혹은 `enum.allCases`에 for문을 사용해 구현한 부분이 많았는데, 불필요한 변수사용을 줄이고 그로인한 side effect를 줄이고자 고차함수 forEach문을 사용했습니다.


 ---

 ## 해결이 되지 않은 점
 - changeSignButton기능에 오류가 있습니다~!
     - Step2 리팩토링 과정에서 함께 해결할 예정입니다.
 - Step2에서 추가 테스트를 통해 발견되는 버그가 있을 시 디버깅을 진행할 예정입니다.

 ---

 ## 조언을 얻고 싶은 부분
- 링크드리스트의 노드 클래스로 구현하는 이유
    - 링크드리스트를 구현할때 다른 사람들의 코드를 참고했는데 Node에 해당하는 코드를 class로 해줬는데 class로 해준 이유가 궁금합니다...
- 폴더&파일 정리하기
    - 계산기 프로젝트에서 생성된 파일이 많아 어떻게 정리를 하는게 좋은것인지 모르겠습니다! 혹시 이런 경우 어떻게 폴더정리를 하는게 좋은 정리법인지 조언 부탁드립니다~!

jonghancha committed:
feat: LinkedList에 clear메서드 추가

DrageunKim committed:
refactor: CalculateError-invalidNumber로 변경

jonghancha committed:
refactor: emptyOperands, emptyOperators Error 추가, Formula.result()메서드 수정

DrageunKim committed:
refactor: componetsByOperator메서드 코드 수정

jonghancha committed:
refactor: ExpressionParser.parse()메서드 수정

DrageunKim committed:
refactor: ViewController 코드 추가

jonghancha committed:
refactor: LinkedListNode.deinit() print문 삭제, Unit Tests Case추가

hyunable commented on Calculator/Calculator/Controller/ViewController.swift:
```diff
@@ -7,12 +7,156 @@
 import UIKit
 
 class ViewController: UIViewController {
```
어떤 역활을 하는 VC 인지 리네이밍 해주세요 ㅎㅎ

DrageunKim replied:
`CalculatorViewController` 로 변경하겠습니다~! ㅎㅎ

hyunable commented:
첫번째 PR 고생많으셨어요 🎉

두 코드를 병합할때 어떤 기준으로 논의가 되었으며 스토리보드 병합시에는 이슈는 없었는지 궁금하네요 ㅎㅎ

협업에서도 두 코드가 병합되면서 머지의 기준에 대해서 논의하는 케이스도 많고, 스토리보드를 사용하는 팀은 컨플릭도 어마어마하기 때문에 이런 경험은 매우 중요한 것 같아요



## QnA

> 링크드리스트의 노드 클래스로 구현하는 이유

-> node는 데이터와 앞 노드를 가리키는 포인터로 이루어져 있습니다. 

링크드리스트에서 노드를 사용할 때, 노드가 struct라면 앞 노드를 가리킬때 어떻게 될까요?



> 폴더&파일 정리하기

-> 요 부분은 같이 협업하는 (코드를 같이 작성하는) 팀원들의 스타일과 팀의 코드 스타일에 따라 매우 다릅니다 ㅎㅎ

아직 그룹핑하는것에 익숙하지 않아서 좋은 사례가 궁금하시다면 스타수가 1k 이상의 github 레포들을 한번 살펴봐보면 많은 도움이 됩니다.

jonghancha commented:
> 링크드리스트에서 노드를 사용할 때, 노드가 struct라면 앞 노드를 가리킬때 어떻게 될까요?

- struct인 노드가 앞 노드를 가리킬 때, 앞 노드의 속성들을 포함한 모든 걸 값타입으로 가지고 있어야 합니다. 구글링과 야곰에게도 질문해본 결과, struct로 구현할 경우 노드라는 객체 안에 무한하게 다음노드들을 포함하고 있어야 하기 때문에 메모리 공간을 예측할 수 없게 된다는 걸 알게 되었습니다. 반면 class로 구현할 경우 previous, next 변수에 노드의 주소값만 저장되기 때문에 컴파일러가 메모리 사용량을 예측할 수 있다고 생각합니다. 

> 두 코드를 병합할때 어떤 기준으로 논의가 되었으며 스토리보드 병합시에는 이슈는 없었는지 궁금하네요 ㅎㅎ

- 계산기 구현코드에서 서로 부족한 부분을 보완하는 방향으로 코드를 합쳤습니다. 기본적으로 정해진 UML의 조건에 맞게 코드를  구현하고자 했고, 코드의 가독성을 높이고 side effect 방지를 위해 고차함수를 적극 사용해보려 했습니다. 스토리보드는 제가 이전 step3 구현을 완료하지 못해 dragon의 코드를 사용했습니다.

DrageunKim committed:
refactor: changeSign기능 관련된 코드 수정

jonghancha committed:
refactor: reset관련 코드 수정

DrageunKim committed:
refactor: updateCalculatorScrollView메서드 추가

DrageunKim committed:
refactor: updateCalculatorScrollView메서드 코드 수정

jonghancha committed:
refactor: 초기값 부호 바꾸기가능하게 수정, showErrorMessage함수 private화

DrageunKim committed:
refactor: changeSign-곱셈,나눗셈경우 개선

DrageunKim committed:
refactor: showResult메서드 코드 수정

jonghancha committed:
refactor: operator나
dquote> =버튼 눌렀을 때 stackCalculation이 FinalCalculation에 추가되도록 수정, CalculatorViewController로 클래스명 수

DrageunKim committed:
refactor: updateCalculatorScrollView() -> scrollToCalculatorScrollViewBottom()으로 메서드명 수정, changeOperator()메서드 추가, pushOperator() 코드 수정

jonghancha committed:
refactor: operator button 통합

DrageunKim committed:
refactor: 곱셈,나눗셈에서 발생하는 에러 개선을 위한 코드 수정

hyunable commented:
넵 맞습니다 ㅎㅎ 메모리 때문이에요! 사소한거라도 이해가 안되는 부분은 의문을 갖고 접근하시면 좋아요 (나중에 어디에선가 터진답니다 ㅠ)
알려고 하는 에티튜드가 좋으세요~!
스토리보드를 `dragon` 의 코드로 사용해서 충돌이 없었군요 ㅎㅎ

제 충돌 경험에 의하면, 스토리보드에 많은 설정을 할 수 록 컨플릭이 많이 나게되고, 코드가 아닌 xml형태의 스토리보드의 컨플릭 코드를 수정하려면 굉장히 오래걸리고 힘들어요 ㅠㅠ(가독성이..) 그렇기 때문에 개인적으로는
스토리보드에는 최소한의 레이아웃만 하고 컴포넌트 설정(backgroundColor, font 등등..)들은 모두 코드로 해주는 편이에요 ㅎㅎ 충돌이 나더라도 코드에서 나는 것이 해결하는데 비용이 덜 들고, 리스크가 적습니다

수고하셨어요 Step2로 고고 해보실까요? ㅎㅎ
